core/arch/arm/kernel/thread_spmc.c

```diff
@@ -1482,13 +1513,33 @@ struct mobj_ffa *thread_spmc_populate_mobj_from_rx(uint64_t cookie)
 	 * OP-TEE is only supporting a single mem_region while the
 	 * specification allows for more than one.
 	 */
-	retrieve_desc = spmc_retrieve_req(cookie);
-	if (!retrieve_desc) {
+	rx_buffer = spmc_retrieve_req(cookie, &length, &fragment_length);
+	if (!rx_buffer) {
 		EMSG("Failed to retrieve cookie from rx buffer %#"PRIx64,
 		     cookie);
 		return NULL;
 	}
 
+	/* Allocate the enough size buffer to copy the fragment descriptor*/
+	retrieve_desc = aligned_alloc(SMALL_PAGE_SIZE, SMALL_PAGE_SIZE * 16);
```

Comment:
We should not use a temporary buffer. We should instead consume each fragment as it is received like in `handle_mem_frag_tx()`. For now, I believe we can assume that each fragment ends with a complete `struct ffa_address_range` like we do in `add_mem_share_helper()`.

Reply:
Sorry, I'm not quite clear what you mean by your last sentence. Could you help me revise it？

Reply:
In `add_mem_share_helper()` we just assume that the buffer starts at a `struct ffa_address_range`. I guess it wouldn't hurt to check that `flen` is a multiple of `sizeof(struct ffa_address_range)`, but the point is that we assume that the SPMC isn't trying to make it harder for us than necessary.

Reply:
Done